### ISSUE TITLE
Screen sharing on mobile app does not exist

### ISSUE THREAD

**Issue by quantumbeat** [jitsi/jitsi-meet#5252]
**Is your feature request related to a problem you are facing?**
Teacher wants to share screen with class
Uses Apple Ipad
Starts Jitsi in Browser
App comes up, can not be avoided
App does not allow to share screen
Materials not visible to students

**Describe the solution you'd like**
Enable screen sharing just as in browser

**Describe alternatives you've considered**
Use desktop computer instead of ipad
But is not portable and has other apps / entire workflow would have to be changed

Thanks for consideration.

**Comment by robbi5** [comment]
I've just looked a bit around what steps are needed for jitsi meet to implement screen sharing from iOS:

**Screen Sharing with Jitsi Meet running in a browser (safari)**
This is currently not possible. Safari for iOS lacks the `getDisplayMedia` API to capture the screens content. The corresponding webkit bug is https://bugs.webkit.org/show_bug.cgi?id=186294

**Screen Sharing with the Jitsi Meet iOS App**
This could work - iOS Apps can access the Screen Content with the [ReplayKit Framework](https://developer.apple.com/documentation/replaykit). For streaming the Screen Content - not only of the Jitsi Meet app itself but from the whole system - into the Jitsi video call, the [ReplayKit2 Broadcast Functionality](https://developer.apple.com/videos/play/wwdc2018/601/) could be used. The Jitsi Meet App for iOS is based on React Native, the only [ReplayKit React Native Module](https://github.com/code-matt/react-native-replaykit) I've found only supports recording, but not the required broadcast functionality.
→ Jitsi Meet for iOS has to implement the ReplayKit2 Broadcast WebRTC Adapter in the native part of the app or extend and implement the react native module for it.

**Comment by bdel** [comment]
Will Jitsi support ReplayKit in future update by luck? Looking forward to it!!

**Comment by JJ-OpusVL** [comment]
There's an article at https://medium.com/better-programming/building-an-ios-screen-recorder-with-react-native-9e8c764c477e which covers processing broadcasts from React Native.

**Comment by martax1970** [comment]
Is ir possible to share the screen using Jitsi on iOS ipad?

**Comment by Filippo-94** [comment]
It is not possible. Skype and others allow it.

**Comment by AliKarpuzoglu** [comment]
@emcho @saghul  Is this feature currently being developed? I want to try and work on it, but if anyone already has any progress, maybe we can continue from there or work on it together

**Comment by mxjoe** [comment]
+1 for Screen Sharing

**Comment by Lakshman1996** [comment]
@robbi5 I was able to implement Broadcast in the native layer with the help of this [article](https://medium.com/better-programming/building-an-ios-screen-recorder-with-react-native-9e8c764c477e), I am getting [sampleBuffer](https://developer.apple.com/documentation/replaykit/rpbroadcastsamplehandler/2123045-processsamplebuffer). Do you have any idea how to push this buffer into Jitsi video feed?

Thanks in advance

**Comment by saghul** [comment]
When I attempted this in the past the challenge was to pass the samples from the extension proess to the main app so we could a video soure with them. Are you at that stage already?

**Comment by AliKarpuzoglu** [comment]
the guy from the article (https://medium.com/better-programming/building-an-ios-screen-recorder-with-react-native-9e8c764c477e ) offered his source code as well:  https://github.com/linuxpi/ScreenRecordingRNDemo

**Comment by Lakshman1996** [comment]
@saghul You're talking about passing the samples from the extension app to react app right? I wasn't sure if i was going in the right direction hence haven't started that yet. If that is the case i can try doing that. Can you let me know what needs to be done post that?

**Comment by ChrisTomAlx** [comment]
@saghul what is the file that is being read here :- https://github.com/jitsi/webrtc/blob/M75/sdk/objc/components/capturer/RTCFileVideoCapturer.m

Is the camera video feed being put into a file and then it is picked up by the capturer ? Is that assumption correct? 
If so then is it naive to think that ios screen sharing would be as simple as getting the buffer from replayKit and then maybe calling a similar function that does not have to read from file and convert to buffer but can directly call the readNextBuffer function implemented in the same file I linked above.

**Comment by saghul** [comment]
> Is the camera video feed being put into a file and then it is picked up by the capturer ? Is that assumption correct?

No. Another capturer is implemented for that. I mentioned the file capturer as an example of another capturer that doesn't use a the camera.

The capturer that takes frames from ReplayKit would need to implement the RTCVideoCapturer API.

**Comment by ChrisTomAlx** [comment]
Thanks a lot for your quick replies.
Okay so this is my understanding

1. Create a button on the React Native APP UI
2. On click of this button stop the camera feed (requires research) and call the implementation of RTCVideoCapturer API that will be created in the the [webrtc React Native Module](https://github.com/react-native-webrtc/react-native-webrtc).
3. Add the broadcast upload extension to the app and send the CMSampleBufferRef buffers obtained from the extension to the [webrtc React Native](https://github.com/react-native-webrtc/react-native-webrtc) module's implementation of the RTCVideoCapturer API. 
4. The RTCVideoCapturer API will handle sending the CMSampleBufferRef obtained from replayKit to the other users and all that is required from us would be to implement a function similar to [publishSampleBuffer](https://github.com/jitsi/webrtc/blob/0cd6ce4de669bed94ba47b88cb71b9be0341bb81/sdk/objc/components/capturer/RTCFileVideoCapturer.m#L157) as seen in the the RTCFileVideoCapturer.m file
5. I have made an assumption that just by implementing the RTCVideoCapturer API  and calling a function similar to [publishSampleBuffer](https://github.com/jitsi/webrtc/blob/0cd6ce4de669bed94ba47b88cb71b9be0341bb81/sdk/objc/components/capturer/RTCFileVideoCapturer.m#L157) we have already pushed our video frames for the other users in the conference to view it. 
6. On click of button again, get the original camera feed if camera was on.

let me know if this just sounds stupid and I am completely on the wrong track. Please suggest if there are any changes to this flow as well.

Thanks a ton for your quick replies!!

**Comment by ChrisTomAlx** [comment]
Hey @saghul 
I see two webrtc repos. 
 1. Branch M75 which I assume is latest on https://github.com/jitsi/webrtc/tree/M75 and 
 2. https://github.com/react-native-webrtc/react-native-webrtc

Would you be so kind as to elaborate on the differences between the two ? 
I see that https://github.com/react-native-webrtc/react-native-webrtc has been added to the package.json here https://github.com/jitsi/jitsi-meet.

So if I were to raise PR for this which repo do I target? Or is 2 somehow a dependency of 1 ?

**Comment by saghul** [comment]
> let me know if this just sounds stupid and I am completely on the wrong track. Please suggest if there are any changes to this flow as well.

That sounds correct.



> I see two webrtc repos.
> 
>     1. Branch M75 which I assume is latest on https://github.com/jitsi/webrtc/tree/M75 and
> 
>     2. https://github.com/react-native-webrtc/react-native-webrtc
> 
> 
> Would you be so kind as to elaborate on the differences between the two ?

1. Is the acctual WebRTC library, you don't need to touch that one.
2. Is the RN wrapper for 1. This is where the new capturer must live. An API would need to be provided to make sure we can "inject" frames from outside, and only enable the capturer if an injector has been attached / enabled.

In addition the sscreen-sharing stream should be exposed with the standard getDisplayMedia API, which would create a stream with a track that uses the screen capturer as its source, if available.

**Comment by ChrisTomAlx** [comment]
Hey @saghul 
Kinda stuck on this. Have a couple of questions.
1. Could you give me a starting point on how react app and react native implements the camera capturer, so I can learn and implement the screen capturer similarly.
2. Have you gotten to the part of adding the react native module as an embedded framework to the broadcast extension? Is this approach even correct ? 

Would appreciate your guidance. Thank you!!

**Comment by Lakshman1996** [comment]
@saghul Getting the CMSampleBuffer from the extension to the app seems to be a near impossible task. So now i am trying to add the JitsiMeet sdk and webrtc framework into the extension and try to join as a new user but with only Video frames (no audio). Is this a viable approach. ?

Thanks

**Comment by saghul** [comment]
@Lakshman1996 I haven't tried but I don't think it's a viiable approach due to the tight memory constraints that exteensions have.

**Comment by Lakshman1996** [comment]
https://github.com/twilio/video-quickstart-ios

Twilio is also doing something similar hence thought might try this way.

**Comment by AliKarpuzoglu** [comment]
@Lakshman1996 @ChrisTomAlx  do you have any code snippets that you could share?

**Comment by ChrisTomAlx** [comment]
@AliKarpuzoglu 
I don't have any code snippets as such, but just some concepts.
On iOS, according to me there are two way to achieve this. Once the broadcast extension is added to the app we get the CMSampleBuffers after that two things can be done :-

1. Either process the CMSampleBuffer within the extension completely. Couldn't really get too far using this method but like saghul said since extension has memory limits it would probably not work as well as expected.
2. Pass the CMSampleBuffer to the app and then follow the same flow as the camera feed does. There are couple of ways I found we could do this. 
     - Use https://github.com/mutualmobile/MMWormhole to communicate small back and forth information if required ( But it cannot be used to send CMSampleBuffer, atleast as far as I know )
     - Use NSFileCoordinator and coordinate read write into the same file from both extension and App. This could work but it might lead to frame drops or lags. since the read write has to happen at a pretty high speed of atleast one read write per 200 - 400 ms.
     - Use Core Data ( Haven't gotten too much into this )

Sorry that's all I can offer right now. Have paused the work on this for now. But I hope these concepts help someone looking into this.

Thanks to all for helping

**Comment by AliKarpuzoglu** [comment]
Hi everyone. 
@saghul 

We are pretty far with the screen sharing but we do require some assistance. Our current solution gets the Screen content, and we can bring it to Swift/React Native. We do, however, struggle to push the content into webrtc... Our current solution manages to share a single frame and then drop.
Is anyone able to help, here's some snippets

fetching video track where we want to write the frame
`  RTCVideoCapturer *videoCapturer = [[RTCVideoCapturer alloc] init];
  RTCMediaStream *stream = self.localStreams[mediaStreamId];
  RTCVideoTrack *videoTrack = stream.videoTracks[0];`

here we write the frame on the video track
`  @try {
           [[videoTrack source] capturer:videoCapturer didCaptureVideoFrame:videoFrame];
        }
        @catch (NSException *exception) {
           NSLog(@"error while writing frame");
        }
`

after one frame, the stream freezes, even though it still is recording on the iOS side

**Comment by arararar** [comment]
We desperately NEED this screencast Feature as teachers, using ipads. PLEASE implement it till the reopening of schools in autumn. Youll get as many coffee as you can use ;)!!!

**Comment by saghul** [comment]
@AliKarpuzoglu Exciting! What thread are you calling those functions in?

**Comment by linuxpi** [comment]
Hi @saghul 

We are executing the following functions on a user initiated dispatch queue:
```
  RTCVideoCapturer *videoCapturer = [[RTCVideoCapturer alloc] init];
  RTCMediaStream *stream = self.localStreams[mediaStreamId];
  RTCVideoTrack *videoTrack = stream.videoTracks[0];
```

and we are writing the frames in a background dispatch queue
```
dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_BACKGROUND, 0), ^{
    while (true) {
      RTCVideoFrame *videoFrame = [SocketShim getNextFrame];
      @try {
           [[videoTrack source] capturer:videoCapturer didCaptureVideoFrame:videoFrame];
        }
        @catch (NSException *exception) {
           NSLog(@"error while writing frame");
        }
    }
});
```

Neither of hem are being executed on the main thread. Let me know if you need any other information or if i missed something here.

**Comment by AliKarpuzoglu** [comment]
@saghul  Screenshare is working. Now our only issue is keeping the app alive in the background. Any ideas?

**Comment by samlll42-github** [comment]
The app should work in the background with the right Audio/VOIP setting. 
https://jitsi.github.io/handbook/docs/dev-guide/dev-guide-ios-sdk
"In order for app to properly work in the background, select the "audio" and "voip" background modes"

Have you tried that? 

We are also working on this - would you mind sharing what you have so we can join forces?

**Comment by jbg** [comment]
@AliKarpuzoglu @linuxpi what is your `[SocketShim getNextFrame]` method doing?

**Comment by anand-hsc** [comment]
> Is anyone able to help, here's some snippets



> Hi everyone.
> @saghul
> 
> We (@linuxpi mostly) are pretty far with the screen sharing but we do require some assistance. Our current solution gets the Screen content, and we can bring it to Swift/React Native. We do, however, struggle to push the content into webrtc... Our current solution manages to share a single frame and then drop.
> Is anyone able to help, here's some snippets
> 
> fetching video track where we want to write the frame
> 
> ```
>   RTCVideoCapturer *videoCapturer = [[RTCVideoCapturer alloc] init];
>   RTCMediaStream *stream = self.localStreams[mediaStreamId];
>   RTCVideoTrack *videoTrack = stream.videoTracks[0];`
> ```
> 
> here we write the frame on the video track
> 
> ```
>   @try {
>            [[videoTrack source] capturer:videoCapturer didCaptureVideoFrame:videoFrame];
>         }
>         @catch (NSException *exception) {
>            NSLog(@"error while writing frame");
>         }
> ```
> 
> after one frame, the stream freezes, even though it still is recording on the iOS side

@AliKarpuzoglu : can u tell us what is meaning of getting screen content to Swift/React native. May be share high-level design approach u followed that will be helpful for us

**Comment by AliKarpuzoglu** [comment]
@saghul it's now running in the background for a longer time (we ran it for 20 mins, unless you open sth like youtube). There are some small issues left, like the button being a bit ugly, but we would love to share it. Are you interested in checking out the testflight, before we place a pull request?

**Comment by saghul** [comment]
@AliKarpuzoglu Sure thing mate! saghul @ jitsi . org Cheers!

**Comment by lagmoellertim** [comment]
Hey @AliKarpuzoglu ,
Would you mind sharing your current state of work? I'm really interested on how you managed to achieve this, I tried to implement it myself, but due to no experience with WebRTC Native, I haven't completed it.

Cheers

**Comment by AliKarpuzoglu** [comment]
@lagmoellertim sure. Send me a mail, it’s on my Website 

Also of antoje Else Wants to join the TestFlight, send me a mail. We’re still working on optimizing it

**Comment by enterux** [comment]
Hey Ali,

Can't find your email, however have sent you dm on twitter. 

Please do respond back on how we can test ?

**Comment by AliKarpuzoglu** [comment]
send an email to testflight (at) alikarpuzoglu (dot) com

**Comment by HoppFoundation** [comment]
https://github.com/HoppFoundation/jitsi-meet/tree/screenshare

Here's the source code. If anyone is able to contribute: 
- Performance could be improved, by sending raw bytes in SampleHandler.swift and receiving them in SocketShim.swift 
- Maybe by improving this performance we don't have to scale down the images as much. However, we need to manage the memory usage 
- we sometimes have an orange screen when sharing, and ending the sharing... Maybe someone can reproduce the error so we find the cause
- (easy) instead of creating random numbers to drop 4/5 Frames just count and take every 5th frame (or other performance improvements)
- localization in the swift part ( the screenshare button and error message)

If you want to try this out, but you don't know how to build it, join the testflight here : https://testflight.apple.com/join/NIyrFWyX

(you'll need to change the server in the settings)

**Comment by joe733** [comment]
> If you want to try this out, but you don't know how to build it, join the testflight here : https://testflight.apple.com/join/NIyrFWyX

Is it ready for android clients too?

**Comment by AliKarpuzoglu** [comment]
> > If you want to try this out, but you don't know how to build it, join the testflight here : https://testflight.apple.com/join/NIyrFWyX
> 
> Is it ready for android clients too?

#5244  already has a working implementation. Maybe there is a way to combine them, but we only worked on iOS

**Comment by xuwab** [comment]
> > > If you want to try this out, but you don't know how to build it, join the testflight here : https://testflight.apple.com/join/NIyrFWyX
> > 
> > 
> > Is it ready for android clients too?
> 
> #5244 already has a working implementation. Maybe there is a way to combine them, but we only worked on iOS

I used xcode 11.6 to compile the framework from screenshare branch. But after integrating into my project, an error was reported:'AppRegistry is not a registered callable module (calling runApplication)'. It is same as issue #7361 .How to solve this problem, bro🙏

**Comment by parveensachdeva** [comment]
Hi @AliKarpuzoglu @HoppFoundation 

I have downloaded the test flight and start broadcasting is fine for me
When I cloned the code, switch to screenshare branch and generated the build with my own bundle Id it is only showing me Start Recording feature. 

I am stuck here Can you please help me

**Comment by AliKarpuzoglu** [comment]
@parveensachdeva 
make sure all the bundle ids are changed... you can find all of them if you  run `grep -rnie 'hopp-foundation' app`  in the ios folder. If that doesn't help try building the Extensions seperately

**Comment by parveensachdeva** [comment]
> @parveensachdeva
> make sure all the bundle ids are changed... you can find all of them if you run `grep -rnie 'hopp-foundation' app` in the ios folder. If that doesn't help try building the Extensions seperately

Thanks @AliKarpuzoglu, for quick reply I will check.

**Comment by alantulais** [comment]
We tested with the provided TestFlight, our comments are: 

1) it looks like it's not posible to share screen if mic was enabled before starting to share screen (app crashes).
2) If you try to play some media (like a song) it stops screen sharing
3) With this code, Android devices can't hangout the meetings.
4) Landscape/Portrait modes are not updated 

Great job!

**Comment by alantulais** [comment]
We tested again and it looks like it is fully working the first time screen sharing is enabled. Second time there are some issues related to playing media (starting screen sharing while mic enabled or playing media like youtube). It thows the following error:

**_Fatal error: Array index is out of range: file_**

@AliKarpuzoglu Any ideas/hints in order to solve this by our side and collaborate to this feature?

**Comment by AliKarpuzoglu** [comment]
Sure, send me an email.
The TestFlight is not the most current version. If you want I’ll add you to it

**Comment by upspire** [comment]
Thanks everyone for your help and support.
Do we have any ETA or update regarding screen-sharing in iOS Apps?

**Comment by saghul** [comment]
No ETA, but we've started looking into the implementation @AliKarpuzoglu / @HoppFoundation and it goes exactly in the direction I was hoping for.

**Comment by saghul** [comment]
> @saghul how would we create a track that doesn't compress the screen the way a video would be compressed? Screenshare on the pc is usually very high quality or not sent at all. I think we might be using a wrong track? Do you think this needs to be added to react-native-webrtc?
> 
> I feel like the screen sharing display looks worse on other iOS devices compared to the desktop.
> For example: Sharing a laptop screen (landscape) and receiving it on a phone (landscape) gives you a higher quality image compared to the phone in portrait...

Sorry I missed this @AliKarpuzoglu . I think the problem is that on Android the VideoSource API is able to mark a track as screen-sharing, but that API seems missing from the ObjC counterpart.

https://github.com/jitsi/webrtc/blob/dc40d5cc81e8fe9aa1cd78a38ee8bb9e91ec49a0/sdk/android/api/org/webrtc/PeerConnectionFactory.java#L463

https://github.com/jitsi/webrtc/blob/dc40d5cc81e8fe9aa1cd78a38ee8bb9e91ec49a0/sdk/objc/api/peerconnection/RTCPeerConnectionFactory.mm#L270

I think the engine would apply the screen-share behavior if this flag was set.

**Comment by AliKarpuzoglu** [comment]
Thats great news @saghul . Will you guys handle the addition to webrtc? 

We have already adjusted the quality  settings quite a bit, and I think with stable internet the quality is already satisfying while not using too much data..  One note: I will try to test it on iPad Pros this week, since they have a higher resolution and we have encountered some issues.  
Looking forward to the results

**Comment by saghul** [comment]
> Will you guys handle the addition to webrtc?

You mean the isScreencast flag on ObjC? I think that's attainable, yeah.

We can hopefully begin working on this in a few week's time.

**Comment by alantulais** [comment]
@saghul @AliKarpuzoglu Hi guys.

We noticed that "Screen share" button was designed in SwiftUI instead of React Native, the last one only receives "onStart" and "onEnd" as callback props.  Is there a reason why it developed like this? If we want to change the implementation to draw the button from React Native with the legends of "Start sharing" and "Stop sharing", what would be the steps to follow and the difficulties we would encounter?

Thank you very much.
Regards.
-Alan

**Comment by saghul** [comment]
Adding React Native to the extension used for the screen-sharing part is probably not worth it.

**Comment by AliKarpuzoglu** [comment]
@saghul am I understanding correctly that we only have to change one bit on the webrtc side to specify that it's a screenshare stream? https://github.com/pristineio/webrtc-mirror/blob/7a5bcdffaab90a05bc1146b2b1ea71c004e54d71/webrtc/api/video/video_content_type.cc#L82 

Similar to how it's done here?

**Comment by J0n4t4n** [comment]
Just wanted to note my support for an implementation of this.
I am currently primarily working from my iPad and I would love to be able to share my iPad screen with others, like I can in Microsoft Teams.
I sadly have no development experience with iOS, but I am more than willing to help test a potential beta version of this feature!

**Comment by saghul** [comment]
> @saghul am I understanding correctly that we only have to change one bit on the webrtc side to specify that it's a screenshare stream? https://github.com/pristineio/webrtc-mirror/blob/7a5bcdffaab90a05bc1146b2b1ea71c004e54d71/webrtc/api/video/video_content_type.cc#L82
> 
> Similar to how it's done here?

@AliKarpuzoglu Yep, that should be it.

**Comment by Rajkiran93** [comment]
@AliKarpuzoglu, @HoppFoundation  and @saghul , could you please explain me on how to integrate the screen sharing functionality. I went through the HoppFoundation repo, I couldn't properly understand how to integrate it with my app.(fyi: I'm a beginner in react native). It's pretty important for me because I have a deadline.
Regards.....

**Comment by saghul** [comment]
Quick update: Android is now integrated and we released a beta with it. A stable release will follow probably this week. iOS is still work in progress.

**Comment by AliKarpuzoglu** [comment]
let me know if we can help in any way @saghul

**Comment by abnerh69** [comment]
Hi @saghul, is it possible to share latest update link to point Android Screen Sharing integrated?  Thanks in advance.

**Comment by saghul** [comment]
The latest release in the Play Store and F-Droid contains screen-sharing. It's version 20.5.0.

**Comment by abnerh69** [comment]
Thank you @saghul, I was wondering if there is source code link available.  Is it open source?

**Comment by saghul** [comment]
Sure thing, this very repository, though the heavy lifting is done in https://github.com/react-native-webrtc/react-native-webrtc/commit/b14787c84670b46cc171126548abd1411a77e8ef

**Comment by alantulais** [comment]
Good morning.

I know that iOS is still in progress, but I have one question. We are trying @AliKarpuzoglu implementation but we are facing some issues related to frame updates. As we know, ReplayKit just send a new frame if the screen has changed. If the user is sharing a presentation and the screen remains static for some seconds, the other users (it is more evident in the other mobile clients) just see the laptop icon while the participants status appears as "Lost". I know that happens when jvb determines that a users must be sending its streaming but it detects no changes after some time. How do you think this can be handled ? Thank you!!

Note: This only happens in lanscape mode because in portrait mode there is a red indicator that causes frequent screen updates.

**Comment by saghul** [comment]
Interesting observation, thanks for sharing!

**Comment by mikagamo** [comment]
Merry Christmas everyone! There is something new for IOS development. Any estimated date for the trial version?

**Comment by 1Ajay** [comment]
Can any one share Commit for android screen sharing

**Comment by sachinhunasnale** [comment]
> https://github.com/HoppFoundation/jitsi-meet/tree/screenshare
> 
> Here's the source code. If anyone is able to contribute:
> 
> * Performance could be improved, by sending raw bytes in SampleHandler.swift and receiving them in SocketShim.swift
> * Maybe by improving this performance we don't have to scale down the images as much. However, we need to manage the memory usage
> * we sometimes have an orange screen when sharing, and ending the sharing... Maybe someone can reproduce the error so we find the cause
> * (easy) instead of creating random numbers to drop 4/5 Frames just count and take every 5th frame (or other performance improvements)
> * localization in the swift part ( the screenshare button and error message)
> 
> If you want to try this out, but you don't know how to build it, join the testflight here : https://testflight.apple.com/join/NIyrFWyX
> 
> (you'll need to change the server in the settings)

if we use socket to write buffer data, do we still need the following files in ur code(ios).
**CustomRTCVideoCapturer.m
WebRTCModuleBridge.m**

**we already have sockets in our code, so can you provide the stpes we follow to capture the recording and share via socket.**

**my assumptions are.
1. implement Screen Recorder with Extension.
2. write the buffer data via socket.**

if we are writing data with socket, should we change anything on server side to indicate that the data is for screen sharing, so that all the participants can see the screen share.

don't know i am right or wrong, but this is how i am planning to implement it.
anyone please provide your feedback on above logic which i am trying to use. should i go with it.
Thank You.

**Comment by saghul** [comment]
We are already close to having it integrated in react-native-webrtc + Jitsi Meet.

**Comment by sachinhunasnale** [comment]
Hi saghul, thanks for quick reply. ll wait for your implementation. the steps i mentioned above are those correct?

**Comment by mikagamo** [comment]
> We are already close to having it integrated in react-native-webrtc + Jitsi Meet.

Do you have any date estimates to test in IOS beta?

**Comment by saghul** [comment]
Please stop asking. Stressing developers out will not yield faster results. It will be ready when it's ready.

**Comment by mikagamo** [comment]
> Please stop asking. Stressing developers out will not yield faster results. It will be ready when it's ready.

Hi Saul,
sorry, it was not my intention to cause stress to anyone. I think you are doing an excellent job.
Thanks

**Comment by Yarkhan** [comment]
iOS screenshare support was merged into master https://github.com/jitsi/jitsi-meet/pull/8705

**Comment by metawops** [comment]
> iOS screenshare support was merged into master #8705

Awesome! Can‘t wait for a new beta build in TestFlight! 👍

**Comment by saghul** [comment]
21.0.0 beta is out. Don't tell anyone ;-)

**Comment by quantumbeat** [comment]
> iOS screenshare support was merged into master #8705

Perfect start into the weekend.
Can't wait to test
A warm and grateful "thank you" to everyone who contributed!

**Comment by amostech** [comment]
Hi @saghul I have a demo for a customer on March 17th 2021. This is a required feature by them. I've checked and they're ok with demonstrating the functionality in Testflight. Is there a way for me to access the version 21.0.0 Beta for iOS? This will mean the world to us! Incredible news.

**Comment by saghul** [comment]
Sure thing, we have the links on the README.

**Comment by tj0xin** [comment]
Hi @saghul this is a great news, while i used the latest master it doesn't show the option to start screen sharing on ios mobile app? is it just me?

**Comment by saghul** [comment]
What iOS version are you using? The feature is available only on iOS >= 14.

**Comment by AliKarpuzoglu** [comment]
I can see the option to screenshare but it shows : 

`org.jitsi.meet.TOGGLE_SCREEN_SHARE` with no listeners registered.`

have you ever seen this saghul?

**Comment by amostech** [comment]
@saghul I found it in the README.md, downloaded, tested and confirmed it works. This is absolutely insane! If we get this contract I need to buy you guys a beer. I love the Jitsi community. I need to get more actively involved with you guys.

**Comment by tj0xin** [comment]
Hi @saghul oops Sorry my bad! it works, i was using iOS 13.5. Now i have updated my test devices to iOS 14.4.1 it shows the option but it wont broadcast. 

My test devices were iPhone 8 Plus and iPad 5th Gen. both on iOS 14.4.1. 

Testflight app works as expected but when i build it won't.

**Comment by saghul** [comment]
> Hi @saghul oops Sorry my bad! it works, i was using iOS 13.5. Now i have updated my test devices to iOS 14.4.1 it shows the option but it wont broadcast. 
> 
> 
> 
> My test devices were iPhone 8 Plus and iPad 5th Gen. both on iOS 14.4.1. 
> 
> 
> 
> Testflight app works as expected but when i build it won't.
> 
> 
> 
> > What iOS version are you using? The feature is available only on iOS >= 14.
> 
> 
> 
> 

You might be missing the app group, did you set that up?

**Comment by saghul** [comment]
> I can see the option to screenshare but it shows : 
> 
> 
> 
> `org.jitsi.meet.TOGGLE_SCREEN_SHARE` with no listeners registered.`
> 
> 
> 
> have you ever seen this saghul?

Where do you see that?

**Comment by AliKarpuzoglu** [comment]
> > I can see the option to screenshare but it shows :
> > `org.jitsi.meet.TOGGLE_SCREEN_SHARE` with no listeners registered.`
> > have you ever seen this saghul?
> 
> Where do you see that?

In the Xcode console, when I  click screenshare and select the extension

**Comment by tj0xin** [comment]
> > Hi @saghul oops Sorry my bad! it works, i was using iOS 13.5. Now i have updated my test devices to iOS 14.4.1 it shows the option but it wont broadcast.
> > My test devices were iPhone 8 Plus and iPad 5th Gen. both on iOS 14.4.1.
> > Testflight app works as expected but when i build it won't.
> > > What iOS version are you using? The feature is available only on iOS >= 14.
> 
> You might be missing the app group, did you set that up?

Yes i did, and i have also changed it in (RTCAppGroupIdentifier) still it is not working

**Comment by tj0xin** [comment]
> > > Hi @saghul oops Sorry my bad! it works, i was using iOS 13.5. Now i have updated my test devices to iOS 14.4.1 it shows the option but it wont broadcast.
> > > My test devices were iPhone 8 Plus and iPad 5th Gen. both on iOS 14.4.1.
> > > Testflight app works as expected but when i build it won't.
> > > > What iOS version are you using? The feature is available only on iOS >= 14.
> > 
> > 
> > You might be missing the app group, did you set that up?
> 
> Yes i did, and i have also changed it in (RTCAppGroupIdentifier) still it is not working

Hi, did anyone succeed with building the app from the latest master with screen sharing working? 

I just tested today it is still not working.

**Comment by alexbumbu** [comment]
> > > > Hi @saghul oops Sorry my bad! it works, i was using iOS 13.5. Now i have updated my test devices to iOS 14.4.1 it shows the option but it wont broadcast.
> > > > My test devices were iPhone 8 Plus and iPad 5th Gen. both on iOS 14.4.1.
> > > > Testflight app works as expected but when i build it won't.
> > > > > What iOS version are you using? The feature is available only on iOS >= 14.
> > > 
> > > 
> > > You might be missing the app group, did you set that up?
> > 
> > 
> > Yes i did, and i have also changed it in (RTCAppGroupIdentifier) still it is not working
> 
> Hi, did anyone succeed with building the app from the latest master with screen sharing working?
> 
> I just tested today it is still not working.

Hi, after tapping the Start screen sharing option is the RPSystemBroadcastPickerView presented?

**Comment by saghul** [comment]
> > > I can see the option to screenshare but it shows :
> > > `org.jitsi.meet.TOGGLE_SCREEN_SHARE` with no listeners registered.`
> > > have you ever seen this saghul?
> > 
> > 
> > Where do you see that?
> 
> In the Xcode console, when I click screenshare and select the extension

Weird, I cannot reproduce that. Moreover, that action is the native one, which is handled by default. Any chance you have an early exception which prevented the listener from being registered?

**Comment by AliKarpuzoglu** [comment]
> > > > I can see the option to screenshare but it shows :
> > > > `org.jitsi.meet.TOGGLE_SCREEN_SHARE` with no listeners registered.`
> > > > have you ever seen this saghul?
> > > 
> > > 
> > > Where do you see that?
> > 
> > 
> > In the Xcode console, when I click screenshare and select the extension
> 
> Weird, I cannot reproduce that. Moreover, that action is the native one, which is handled by default. Any chance you have an early exception which prevented the listener from being registered?

Hey,
So i got it working now. I am sadly not sure what exactly caused it, I think it was a bundle identifier mismatch at some place.

**Comment by saghul** [comment]
Thanks @AliKarpuzoglu glad that got solved!

**Comment by tj0xin** [comment]
> > > > > Hi @saghul oops Sorry my bad! it works, i was using iOS 13.5. Now i have updated my test devices to iOS 14.4.1 it shows the option but it wont broadcast.
> > > > > My test devices were iPhone 8 Plus and iPad 5th Gen. both on iOS 14.4.1.
> > > > > Testflight app works as expected but when i build it won't.
> > > > > > What iOS version are you using? The feature is available only on iOS >= 14.
> > > > 
> > > > 
> > > > You might be missing the app group, did you set that up?
> > > 
> > > 
> > > Yes i did, and i have also changed it in (RTCAppGroupIdentifier) still it is not working
> > 
> > 
> > Hi, did anyone succeed with building the app from the latest master with screen sharing working?
> > I just tested today it is still not working.
> 
> Hi, after tapping the Start screen sharing option is the RPSystemBroadcastPickerView presented?

Hey, 
yes it is presented and i can select that as well, its just that it doesn't start broadcast and then again if you look in More options the option is reset to Start screen sharing.

**Comment by saghul** [comment]
Do you see the Jitsi extension in the list?

**Comment by tj0xin** [comment]
> Do you see the Jitsi extension in the list?

Yes i do

**Comment by saghul** [comment]
Did you put your group ID here as well as in the entitlements file? https://github.com/jitsi/jitsi-meet/blob/f3c1b8ac088cda7c543c04a2306a6a4c20841774/ios/app/src/Info.plist#L79

**Comment by tj0xin** [comment]
> Did you put your group ID here as well as in the entitlements file?
> 
> https://github.com/jitsi/jitsi-meet/blob/f3c1b8ac088cda7c543c04a2306a6a4c20841774/ios/app/src/Info.plist#L79

Yes i have done that.

**Comment by alexbumbu** [comment]
Did you update the `appGroupIdentifier` value in the SampleHandler?

**Comment by tj0xin** [comment]
> Did you update the `appGroupIdentifier` value in the `SampleHandler`?

Yes i did update

**Comment by AliKarpuzoglu** [comment]
@saghul do you know if the servers need some update to support the screen sharing? Screensharing doesn’t work on our self hosted instance, but it does on meet.jit.si

**Comment by saghul** [comment]
No, there shouldn't be any change necessary.

Are you running into this issue in P2P mode or also with more than 3 participants?

**Comment by tj0xin** [comment]
@saghul @alexbumbu 

Thanks, i started over from scratch and now it is working.

**Comment by saghul** [comment]
Excellent!

**Comment by tj0xin** [comment]
Does anyone have this error while distributing app? Having searched a lot it looks like i have to create a new app and bundle ID. 

ERROR ITMS-90347: "Bad bundle identifier. The bundle identifier 'com.company.app.broadcast.extension' of the application extension jitsi-meet.app/PlugIns/JitsiMeetBroadcast Extension.appex should start with the application's bundle identifier 'com.company.app' and not contain more than one period “.” after the application's bundle ID."

**Comment by saghul** [comment]
Yeah you need to have them match. See how we do it here: https://github.com/jitsi/jitsi-meet/blob/master/ios/fastlane/Fastfile

**Comment by tj0xin** [comment]
> Yeah you need to have them match. See how we do it here: https://github.com/jitsi/jitsi-meet/blob/master/ios/fastlane/Fastfile

Solved it, just removed .extension after com.company.app.broadcast.extension so now it is com.company.app.broadcast

Apple is getting strict with bundle id extensions!!!

**Comment by AliKarpuzoglu** [comment]
> No, there shouldn't be any change necessary.
> 
> Are you running into this issue in P2P mode or also with more than 3 participants?

I also run into it when there's more than 3 participants

**Comment by saghul** [comment]
Can you set the `capScreenshareBitrate` config option to 1 and test?

**Comment by tj0xin** [comment]
> Can you set the `capScreenshareBitrate` config option to 1 and test?

I also have this problem when there are 3 or more participants, tried capScreenshareBitrate it didn't work.

**Comment by diptendulkar** [comment]
Does it work on the simulator with ios > = 14  or any specific changes required to make it work on simulators

**Comment by saghul** [comment]
> Does it work on the simulator with ios > = 14 or any specific changes required to make it work on simulators

No, it doesn't work on the simulator, you need a real device to test it.

**Comment by eriktor1** [comment]
Same issue with our servers. sharing screen from Jitsi App (or SDK) doesnt work to IoS (p2p), nor to anyone in a 3 party (or more) meeting. but works well with JIT.SI. Very Strange.
thought it might be an issue with protocol handling at the beginning, but couldnt find any related parameter in the config files

**Comment by saghul** [comment]
What version are you running on your servers?

**Comment by eriktor1** [comment]
latest stable (5765)

**Comment by eriktor1** [comment]
more precisely ;-) : 
    jitsi-meet:
      Installed: 2.0.5765-1
    jitsi-meet-prosody:
      Installed: 1.0.4900-1
    jitsi-meet-tokens:
      Installed: 1.0.4900-1
    jitsi-meet-turnserver:
      Installed: 1.0.4900-1
    jicofo:
      Installed: 1.0-740-1
    jitsi-meet-web:
      Installed: 1.0.4900-1
    jitsi-meet-web-config:
      Installed: 1.0.4900-1
    jitsi-videobridge2:
      Installed: 2.1-478-gc6da57bd-1

**Comment by eriktor1** [comment]
seems same issue as in https://community.jitsi.org/t/screen-sharing-from-android-app-to-browser-on-self-hosted-instance-does-not-work/85734
but cannot understand why its working on JIT.SI , but should requiere a fix on React to work on a self hosted server? Make no sense...

**Comment by eriktor1** [comment]
have come to the idea that its maybe related to an issue of how the automated installation process works: maybe a bug in one server deployment, vs a distributed deployment (as in JIT.SI)?

**Comment by swarajitservices15** [comment]
@saghul : I am trying to enable screen sharing option. I have added the latest iOS SDK 3.5.0 (iOS SDK URL ref: https://github.com/jitsi/jitsi-meet-ios-sdk-releases/releases/tag/3.5.0) in my project and checked for screen sharing, but I can not see the screen share option. Please check the screenshot attached.
![IMG_6639](https://user-images.githubusercontent.com/83570442/117124377-e1911680-adb5-11eb-99f5-a607c939e58b.PNG)

Am I missing anything else to enable screen share option? 
Do I need to do any additional code base settings?
Please help me in this.
Thanks in advance.....

**Comment by sagarsukode** [comment]
@saghul : I am trying to enable screen sharing option. I have added the latest iOS SDK 3.5.0 (iOS SDK URL ref: https://github.com/jitsi/jitsi-meet-ios-sdk-releases/releases/tag/3.5.0) in my project and checked for screen sharing, but I can not see the screen share option. Please check the screenshot attached.
![IMG_6639](https://user-images.githubusercontent.com/83570442/117124377-e1911680-adb5-11eb-99f5-a607c939e58b.PNG)

Am I missing anything else to enable screen share option? 
Do I need to do any additional code base settings?
Please help me in this.
Thanks in advance.....

**Comment by saghul** [comment]
You need to provide a broadcast extension and set the iOS screen-sharing flag, it's not as seamless as Android. We haven't documented this properly yet but it's coming. In the meantime you can look at how the Jitsi Meet app itself works.

**Comment by sagarsukode** [comment]
> You need to provide a broadcast extension and set the iOS screen-sharing flag, it's not as seamless as Android. We haven't documented this properly yet but it's coming. In the meantime you can look at how the Jitsi Meet app itself works.

@saghul : Thank you for your quick response. 
Ok, I will try this.
Also, Can you please make some documentation for screen sharing ASAP?

**Comment by Rajkiran93** [comment]
@saghul or anyone tried screen share by integrating 3.5.0? I enabled screen share option, broad cast extension is showing up when I initiate the screen share, app groups are also added but the its not working am I missing anything?

**Comment by maheshk15** [comment]
@saghul screen sharing is working fine for first time when i join the meeting, when i leave the meeting and try to join again jitsi meet failed to create local track audio and video is not working, please let me know if i missing anything?

**Comment by subhash4android** [comment]
On self hosted server, ios screenshare is not working for more than 2 participants.
Did any one tried for more than 2 participants on ios

**Comment by idrive-pankaj** [comment]
@saghul I have the latest code and I have added RTCAppGroupIdentifier in the info.plist file but still screen share option is not coming.
Please let me know if I need to add any other changes.
After creating the group identifier on apple dev account I have updated my profile as well and using the latest one.

**Comment by saghul** [comment]
Please check the documentation in the handbook as well as the sample application.

**Comment by subhash4android** [comment]
> On self hosted server, ios screenshare is not working for more than 2 participants.
> Did any one tried for more than 2 participants on ios

Has anyonce faced this issue?

**Comment by AliKarpuzoglu** [comment]
> > On self hosted server, ios screenshare is not working for more than 2 participants.
> > Did any one tried for more than 2 participants on ios
> 
> Has anyonce faced this issue?

It works for me

**Comment by subhash4android** [comment]
OK ,
For me only one to one participants screen sharing is working. If any 3rd participant joins the meeting its not working ,
showing only screen sharing icon in tile view.

**Comment by xiruoruo** [comment]
> @saghul : I am trying to enable screen sharing option. I have added the latest iOS SDK 3.5.0 (iOS SDK URL ref: https://github.com/jitsi/jitsi-meet-ios-sdk-releases/releases/tag/3.5.0) in my project and checked for screen sharing, but I can not see the screen share option. Please check the screenshot attached.
> ![IMG_6639](https://user-images.githubusercontent.com/83570442/117124377-e1911680-adb5-11eb-99f5-a607c939e58b.PNG)
> 
> Am I missing anything else to enable screen share option?
> Do I need to do any additional code base settings?
> Please help me in this.
> Thanks in advance.....
> **Note: I am checking on iOS device: iPhone 11, iOS version: 14.0**

hello, I encountered the same problem as you. How did you solve it in the end?

**Comment by mikagamo** [comment]
Hi All

Hi,

I have implemented the latest JitsiMeet iOS SDK Pod (3.8.1) in my app and I have followed the guide to allow screen sharing support: iOS SDK · Jitsi Meet Handbook 1

Although the Broadcast Receiver is working and everything seems to be fine, when I start the screen sharing the other participants only see a gray screen. In order to check the implementation I have downloaded the screen sharing sample: GitHub - jitsi/jitsi-meet-sdk-samples: Jitsi Meet SDK examples (Android and iOS)

By default it doesn’t even compile, but after fixing it gives the same result: gray image when sharing the screen. I have tested the meeting in the Jitsi servers and in a custom one but nothing changes. I’m deploying in Xcode 12.5.1 with iOS 14.4 and I have tested it in an iPad with iOS 14.7.1.

Do anyone have experimented this kind of issue or knows a workaround?

Thanks!

Best Reagards

**Comment by blochhamiltonian** [comment]
I’ve heard that now getDisplayMedia API is available on Safari 13 (as this is apparently what Hangouts uses). Is the ability to screenshare on the web browser version of Jitsi in Safari on iOS going to soon be possible? It’d be a massive help. Thank you :)

**Comment by jbg** [comment]
`getDisplayMedia` is supported in macOS Safari 13. It's still not supported at all in iOS Safari.
